src/file_pipe_log/pipe_builder.rs

```diff
@@ -13,17 +14,74 @@ use crate::event_listener::EventListener;
 use crate::file_builder::FileBuilder;
 use crate::log_batch::LogItemBatch;
 use crate::pipe_log::{FileId, FileSeq, LogQueue};
+use crate::truncate::TruncateQueueParameter;
 use crate::Result;
 
 use super::format::FileNameExt;
 use super::log_file::{build_file_reader, LogFd};
 use super::pipe::{DualPipes, SinglePipe};
 use super::reader::LogItemBatchFileReader;
 
-pub trait ReplayMachine: Send + Default {
-    fn replay(&mut self, item_batch: LogItemBatch, file_id: FileId) -> Result<()>;
+pub trait ReplayMachine: Send + Default + Sync {
+    fn replay(&mut self, item_batch: LogItemBatch, file_id: FileId, path: &Path) -> Result<()>;
 
     fn merge(&mut self, rhs: Self, queue: LogQueue) -> Result<()>;
+
+    fn truncate<B: FileBuilder>(&mut self, _builder: Arc<B>) -> Result<()> {
+        Ok(())
+    }
+
+    fn init(&mut self, _params: Option<TruncateQueueParameter>) {}
+}
+
+pub trait ReplayMachineFactory: Sync {
+    type Machine: ReplayMachine;
+
+    fn new_machine(&self) -> Self::Machine;
+
+    fn parameters(&self) -> Option<TruncateQueueParameter>;
```

Comment:
Remove this, internal parameters should be exposed.

Reply:
OK, I have rename the method and return the boolean value that whether we should do truncate operation.